### PR TITLE
LPS-83052 Fix ModuleNameUtil to fit the changes we are doing to semantics

### PR DESCRIPTION
**Pull request by izaera**
This PR uses an LRU cache in the servlet side to avoid introducing breaking changes in the npm API. This is to ease backport but, at the same time, not incurring in big performance penalties.

The cache size is bound to a maximum length equal to the number of resolved JS packages. Because the number of resolved JS packages may grow or shrink with time, the cache will adapt to those changes automatically by enlarging the cache size or by removing stale cache entries.
 
In a future PR we will add a new method in `NPMRegistry` to be able to directly lookup resolved packages in the inner map that `NPMRegistry` holds and thus be able to avoid the LRU cache.

### CODE REVIEW

**Comment by izaera** [comment]
ci:test

File: modules/apps/frontend-js/frontend-js-loader-modules-extender-api/src/main/java/com/liferay/frontend/js/loader/modules/extender/npm/ModuleNameUtil.java [hunk 1]
```diff
@@ -208,16 +207,18 @@ public static String toFileName(String moduleName) {
...
-			fileName = fileName.substring(2);
-		}
-
 		int i = fileName.lastIndexOf(CharPool.PERIOD);
 
 		if (i == -1) {
 			return fileName;
 		}
 
+		String extension = fileName.substring(i);
```
Comment (jbalsas) [inline modules/apps/frontend-js/frontend-js-loader-modules-extender-api/src/main/java/com/liferay/frontend/js/loader/modules/extender/npm/ModuleNameUtil.java]: I think we have `FileUtil.getExtension` for this...

File: modules/apps/frontend-js/frontend-js-loader-modules-extender/src/main/java/com/liferay/frontend/js/loader/modules/extender/internal/npm/builtin/BaseBuiltInJSModuleServlet.java [hunk 2]
```diff
@@ -38,73 +37,66 @@
...
-
-		String identifier = pathInfo.substring(1);
-
-		if (pathInfo.endsWith(".map")) {
-			int index = identifier.lastIndexOf(StringPool.PERIOD);
+	private void _setContentType(HttpServletResponse response, URL url) {
+		String file = url.getFile();
 
-			identifier = identifier.substring(0, index);
+		if (file.endsWith(".js")) {
```
Comment (jbalsas) [inline modules/apps/frontend-js/frontend-js-loader-modules-extender/src/main/java/com/liferay/frontend/js/loader/modules/extender/internal/npm/builtin/BaseBuiltInJSModuleServlet.java]: I think we could also use here `FileUtil.getExtension`, right?

File: modules/apps/frontend-js/frontend-js-loader-modules-extender/src/main/java/com/liferay/frontend/js/loader/modules/extender/internal/npm/builtin/BuiltInJSResolvedModuleServlet.java [hunk 3]
```diff
@@ -37,13 +45,79 @@
...
+		if (jsPackage == null) {
+			return null;
+		}
+
+		String packagePath = ModuleNameUtil.getPackagePath(identifier);
+
+		return jsPackage.getResourceURL(packagePath);
+	}
+
+	private JSPackage _getResolvedJSPackage(String resolvedPackageId) {
```
Comment (jbalsas) [inline modules/apps/frontend-js/frontend-js-loader-modules-extender/src/main/java/com/liferay/frontend/js/loader/modules/extender/internal/npm/builtin/BuiltInJSResolvedModuleServlet.java]: I'd say this probably needs to be... `String resolvedJSPackageId`? 😂

File: modules/apps/frontend-js/frontend-js-loader-modules-extender-api/src/main/java/com/liferay/frontend/js/loader/modules/extender/npm/ModuleNameUtil.java [hunk 4]
```diff
@@ -119,7 +119,6 @@ public static String getModuleId(JSPackage jsPackage, String moduleName) {
 	 * @param  moduleName the module's name
 	 * @return the package name or <code>null</code> if the module name is a
 	 *         reserved or local one
-	 * @review
```
Comment (jbalsas) [inline modules/apps/frontend-js/frontend-js-loader-modules-extender-api/src/main/java/com/liferay/frontend/js/loader/modules/extender/npm/ModuleNameUtil.java]: Not sure if we can remove the `@review comments... this is still pending review by the docs team, I guess :thinking:

File: modules/apps/frontend-js/frontend-js-loader-modules-extender/src/main/java/com/liferay/frontend/js/loader/modules/extender/internal/npm/builtin/BaseBuiltInJSModuleServlet.java [hunk 5]
```diff
@@ -38,73 +37,66 @@
...
+		else if (file.endsWith(".map")) {
+			response.setContentType(ContentTypes.APPLICATION_JSON);
+		}
+		else {
+			MimeTypes mimeTypes = getMimeTypes();
 
-		String moduleName = ModuleNameUtil.toModuleName(identifier);
-
-		return getJSModule(moduleName);
+			response.setContentType(mimeTypes.getContentType(file));
```
Comment (jbalsas) [inline modules/apps/frontend-js/frontend-js-loader-modules-extender/src/main/java/com/liferay/frontend/js/loader/modules/extender/internal/npm/builtin/BaseBuiltInJSModuleServlet.java]: Does this mean we can now serve anything from this servlet?

**Comment by jbalsas** [comment]
Pull request submitted to https://github.com/izaera/liferay-portal/pull/55. [See changes here.](https://github.com/jbalsas/liferay-portal/compare/9abc9a2c792f~1...9abc9a2c792f) <br><br>:octocat: *Sent from [GH](http://nodegh.io).*

**Comment by jbalsas** [comment]
Pull request submitted to https://github.com/izaera/liferay-portal/pull/56. [See changes here.](https://github.com/jbalsas/liferay-portal/compare/1d489d3c35ad~1...1d489d3c35ad) <br><br>:octocat: *Sent from [GH](http://nodegh.io).*

**Pull request closed without merging**